File: backend/services/links.go
package services

import (
	"LinkShorter/models"
	"context"
	"errors"
	"fmt"
	"strings"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

/*
Файл links.go содержит функции для работы со ссылками.
Создание коротких и кастомных ссылок, получение оригинального
URL по короткому коду, получение списка ссылок пользователя
с сортировкой/фильтрацией/поиском, учёт просмотров, работу
с тэгами и удаление ссылок
*/

/*
GenerateLink возвращает короткую ссылку для указанного
оригинального адреса. Если ссылка для этого URL уже
существует (в анонимном пуле при userID = 0 или в личном
пуле пользователя), возвращает существующий короткий код.
Иначе генерирует случайный уникальный код (с проверкой
на коллизии) и сохраняет новую ссылку в базе данных
*/
func GenerateLink(db *pgxpool.Pool, longLink string, userID int) (string, error) {
	var link = "localhost:8080/"
	var existedLink string

	if userID == 0 {
		err := db.QueryRow(
			context.Background(),
			`SELECT short_url FROM links
				WHERE original_url = $1 AND user_id = 0`,
			longLink,
		).Scan(&existedLink)

		if errors.Is(err, pgx.ErrNoRows) {
			var id int
			for {
				var randomEnd = ToUpperAndLower((uuid.New()).String())[0:6]

				err := db.QueryRow(
					context.Background(),
					`SELECT id FROM links
				     WHERE short_url = $1`,
					link+randomEnd,
				).Scan(&id)

				if errors.Is(err, pgx.ErrNoRows) {
					link = link + randomEnd
					_, err = db.Exec(
						context.Background(),
						`INSERT INTO links
						 (original_url, short_url) VALUES ($1, $2)`,
						longLink, link,
					)

					if err != nil {
						return "", err
					}

					return link, nil

				} else if err != nil {
					return "", err
				}
			}
		} else if err != nil {
			return "", err
		}

		return existedLink, nil
	}

	err := db.QueryRow(
		context.Background(),
		`SELECT short_url FROM links
				WHERE original_url = $1 AND user_id = $2 AND is_custom = FALSE`,
		longLink, userID,
	).Scan(&existedLink)

	if errors.Is(err, pgx.ErrNoRows) {
		var id int
		for {
			var randomEnd = ToUpperAndLower((uuid.New()).String())[0:6]

			err := db.QueryRow(
				context.Background(),
				`SELECT id FROM links
				     WHERE short_url = $1`,
				link+randomEnd,
			).Scan(&id)

			if errors.Is(err, pgx.ErrNoRows) {
				link = link + randomEnd
				_, err = db.Exec(
					context.Background(),
					`INSERT INTO links
						 (original_url, short_url, user_id) VALUES ($1, $2, $3)`,
					longLink, link, userID,
				)

				if err != nil {
					return "", err
				}

				return link, nil

			} else if err != nil {
				return "", err
			}
		}
	} else if err != nil {
		return "", err
	}

	return existedLink, nil
}

/*
ReturnOriginalLink возвращает оригинальную ссылку
по коду короткой ссылки
*/
func ReturnOriginalLink(db *pgxpool.Pool, code string) (string, error) {
	var originalURL string
	var shortURL = "localhost:8080/" + code

	err := db.QueryRow(
		context.Background(),
		`SELECT original_url FROM links WHERE short_url = $1`,
		shortURL,
	).Scan(&originalURL)

	if err != nil {
		return "", err
	}

	return originalURL, nil
}

/*
UserLinks возвращает список ссылок пользователя
с учётом поиска, сортировки и выбранных
параметров фильтрации
*/
func UserLinks(db *pgxpool.Pool,
	userId int,
	search string,
	sortBy string,
	orderBy string,
	filterPeriod string,
	filterViews string,
	filterTags string) ([]models.Link, error) {

	var searchQuery = ""
	if strings.TrimSpace(search) != "" {
		searchQuery = "AND tag ILIKE $2"
	}

	var sortColumn string
	switch sortBy {
	case "date":
		sortColumn = "created_at"
	case "views":
		sortColumn = "views"
	case "tag":
		sortColumn = `LOWER(tag) COLLATE "ru-x-icu"`
	default:
		return nil, errors.New("Неверная сортировка")
	}

	var sortOrder string
	switch orderBy {
	case "asc":
		sortOrder = "ASC"
	case "desc":
		sortOrder = "DESC"
	default:
		return nil, errors.New("Неверный порядок сортировки")
	}

	var filterPeriodQuery string
	switch filterPeriod {
	case "all":
		filterPeriodQuery = ""
	case "week":
		filterPeriodQuery = "AND created_at >= NOW() - INTERVAL '7 days'"
	case "month":
		filterPeriodQuery = "AND created_at >= NOW() - INTERVAL '1 month'"
	case "year":
		filterPeriodQuery = "AND created_at >= NOW() - INTERVAL '1 year'"
	default:
		return nil, errors.New("Неверный параметр фильтрации по периоду")
	}

	var filterViewsQuery string
	switch filterViews {
	case "0":
		filterViewsQuery = ""
	case "10":
		filterViewsQuery = "AND views >= 10"
	case "100":
		filterViewsQuery = "AND views >= 100"
	case "1000":
		filterViewsQuery = "AND views >= 1000"
	case "10000":
		filterViewsQuery = "AND views >= 10000"
	default:
		return nil, errors.New("Неверный параметр фильтрации по просмотрам")
	}

	var filterTagsQuery string
	switch filterTags {
	case "all":
		filterTagsQuery = ""
	case "with":
		filterTagsQuery = "AND tag <> ''"
	case "without":
		filterTagsQuery = "AND tag = ''"
	default:
		return nil, errors.New("Неверный параметр фильтрации по тэгам")
	}

	query := fmt.Sprintf(`
		SELECT id,
		       short_url,
		       original_url,
		       views,
		       tag,
		       created_at
		FROM links
		WHERE user_id = $1
		%s
		%s
		%s
		%s
		ORDER BY %s %s
	`, filterPeriodQuery,
		filterViewsQuery,
		filterTagsQuery,
		searchQuery,
		sortColumn, sortOrder)

	var rows pgx.Rows
	var err error
	if searchQuery == "" {
		rows, err = db.Query(
			context.Background(),
			query,
			userId,
		)
	} else {
		rows, err = db.Query(
			context.Background(),
			query,
			userId,
			"%"+search+"%",
		)
	}

	if err != nil {
		return nil, err
	}

	defer rows.Close()

	var links []models.Link
	for rows.Next() {
		var link models.Link
		if err := rows.Scan(&link.LinkID, &link.ShortURL, &link.OriginalURL, &link.Views, &link.Tag, &link.CreatedAt); err != nil {
			return nil, err
		}
		links = append(links, link)
	}

	return links, rows.Err()
}

/*
IncreaseLinkViews увеличивает количество
просмотров короткой ссылки на единицу
*/
func IncreaseLinkViews(db *pgxpool.Pool, code string) error {
	_, err := db.Exec(
		context.Background(),
		`UPDATE links
		SET views = views + 1
		WHERE short_url LIKE '%' || $1 || '%'`,
		code)

	if err != nil {
		return err
	}

	return nil
}

/*
DeleteLink удаляет указанную ссылку пользователя
из базы данных
*/
func DeleteLink(db *pgxpool.Pool, linkID int, userID int) error {
	_, err := db.Exec(
		context.Background(),
		`DELETE FROM links WHERE id=$1 AND user_id=$2`,
		linkID, userID)

	if err != nil {
		return err
	}

	return nil
}

/*
AddTag добавляет или обновляет тег
для указанной ссылки
*/
func AddTag(db *pgxpool.Pool, linkID int, tag string) error {
	_, err := db.Exec(
		context.Background(),
		`UPDATE links SET tag=$1 WHERE id=$2`,
		tag, linkID)

	if err != nil {
		return err
	}

	return nil
}

/*
CreateCustomLink создаёт пользовательскую короткую
ссылку, если выбранное название ещё не занято,
и возвращает результат создания
*/
func CreateCustomLink(db *pgxpool.Pool, userID int, originalLink string, custom string) (string, bool, error) {
	var link = "localhost:8080/" + custom
	var exist bool

	err := db.QueryRow(
		context.Background(),
		`SELECT EXISTS(
			SELECT 1 FROM links
			WHERE short_url=$1
		)`, link).Scan(&exist)

	if err != nil {
		return "", true, err
	}

	if exist {
		return "", true, nil
	}

	_, err = db.Exec(
		context.Background(),
		`INSERT INTO links 
			(user_id, original_url, short_url, is_custom) VALUES ($1, $2, $3, $4)`,
		userID, originalLink, link, true)

	if err != nil {
		return "", true, err
	}

	return link, false, nil
}

File: backend/services/links.rs
use sqlx::PgPool;

use crate::models::Link;
use crate::services::utils::to_upper_and_lower;

const BASE_URL: &str = "localhost:8080/";

fn generate_random_code() -> String {
    let uuid = uuid::Uuid::new_v4().to_string();
    to_upper_and_lower(&uuid)[0..6].to_string()
}

pub async fn generate_link(
    pool: &PgPool,
    long_link: &str,
    user_id: i32,
) -> Result<String, sqlx::Error> {
    if user_id == 0 {
        let existing: Option<(String,)> = sqlx::query_as(
            "SELECT short_url FROM links WHERE original_url = $1 AND user_id = 0",
        )
        .bind(long_link)
        .fetch_optional(pool)
        .await?;

        if let Some((short_url,)) = existing {
            return Ok(short_url);
        }

        loop {
            let candidate = format!("{}{}", BASE_URL, generate_random_code());

            let taken: Option<(i32,)> =
                sqlx::query_as("SELECT id FROM links WHERE short_url = $1")
                    .bind(&candidate)
                    .fetch_optional(pool)
                    .await?;

            if taken.is_none() {
                sqlx::query(
                    "INSERT INTO links (original_url, short_url) VALUES ($1, $2)",
                )
                .bind(long_link)
                .bind(&candidate)
                .execute(pool)
                .await?;

                return Ok(candidate);
            }
        }
    }

    let existing: Option<(String,)> = sqlx::query_as(
        "SELECT short_url FROM links WHERE original_url = $1 AND user_id = $2 AND is_custom = FALSE",
    )
    .bind(long_link)
    .bind(user_id)
    .fetch_optional(pool)
    .await?;

    if let Some((short_url,)) = existing {
        return Ok(short_url);
    }

    loop {
        let candidate = format!("{}{}", BASE_URL, generate_random_code());

        let taken: Option<(i32,)> = sqlx::query_as("SELECT id FROM links WHERE short_url = $1")
            .bind(&candidate)
            .fetch_optional(pool)
            .await?;

        if taken.is_none() {
            sqlx::query(
                "INSERT INTO links (original_url, short_url, user_id) VALUES ($1, $2, $3)",
            )
            .bind(long_link)
            .bind(&candidate)
            .bind(user_id)
            .execute(pool)
            .await?;

            return Ok(candidate);
        }
    }
}

pub async fn return_original_link(pool: &PgPool, code: &str) -> Result<String, sqlx::Error> {
    let short_url = format!("{}{}", BASE_URL, code);

    let row: (String,) = sqlx::query_as("SELECT original_url FROM links WHERE short_url = $1")
        .bind(&short_url)
        .fetch_one(pool)
        .await?;

    Ok(row.0)
}

pub async fn user_links(
    pool: &PgPool,
    user_id: i32,
    search: &str,
    sort_by: &str,
    order_by: &str,
    filter_period: &str,
    filter_views: &str,
    filter_tags: &str,
) -> Result<Vec<Link>, String> {
    let search_trimmed = search.trim();
    let use_search = !search_trimmed.is_empty();

    let sort_column = match sort_by {
        "date" => "created_at",
        "views" => "views",
        "tag" => r#"LOWER(tag) COLLATE "ru-x-icu""#,
        _ => return Err("Неверная сортировка".to_string()),
    };

    let sort_order = match order_by {
        "asc" => "ASC",
        "desc" => "DESC",
        _ => return Err("Неверный порядок сортировки".to_string()),
    };

    let filter_period_query = match filter_period {
        "all" => "",
        "week" => "AND created_at >= NOW() - INTERVAL '7 days'",
        "month" => "AND created_at >= NOW() - INTERVAL '1 month'",
        "year" => "AND created_at >= NOW() - INTERVAL '1 year'",
        _ => return Err("Неверный параметр фильтрации по периоду".to_string()),
    };

    let filter_views_query = match filter_views {
        "0" => "",
        "10" => "AND views >= 10",
        "100" => "AND views >= 100",
        "1000" => "AND views >= 1000",
        "10000" => "AND views >= 10000",
        _ => return Err("Неверный параметр фильтрации по просмотрам".to_string()),
    };

    let filter_tags_query = match filter_tags {
        "all" => "",
        "with" => "AND tag <> ''",
        "without" => "AND tag = ''",
        _ => return Err("Неверный параметр фильтрации по тэгам".to_string()),
    };

    let search_query = if use_search { "AND tag ILIKE $2" } else { "" };

    let query = format!(
        r#"
        SELECT id,
               short_url,
               original_url,
               views,
               tag,
               created_at
        FROM links
        WHERE user_id = $1
        {}
        {}
        {}
        {}
        ORDER BY {} {}
        "#,
        filter_period_query,
        filter_views_query,
        filter_tags_query,
        search_query,
        sort_column,
        sort_order
    );

    let result = if use_search {
        let pattern = format!("%{}%", search_trimmed);
        sqlx::query_as::<_, Link>(&query)
            .bind(user_id)
            .bind(pattern)
            .fetch_all(pool)
            .await
    } else {
        sqlx::query_as::<_, Link>(&query)
            .bind(user_id)
            .fetch_all(pool)
            .await
    };

    result.map_err(|e| e.to_string())
}

pub async fn increase_link_views(pool: &PgPool, code: &str) -> Result<(), sqlx::Error> {
    sqlx::query("UPDATE links SET views = views + 1 WHERE short_url LIKE '%' || $1 || '%'")
        .bind(code)
        .execute(pool)
        .await?;

    Ok(())
}

pub async fn delete_link(pool: &PgPool, link_id: i32, user_id: i32) -> Result<(), sqlx::Error> {
    sqlx::query("DELETE FROM links WHERE id=$1 AND user_id=$2")
        .bind(link_id)
        .bind(user_id)
        .execute(pool)
        .await?;

    Ok(())
}

pub async fn add_tag(pool: &PgPool, link_id: i32, tag: &str) -> Result<(), sqlx::Error> {
    sqlx::query("UPDATE links SET tag=$1 WHERE id=$2")
        .bind(tag)
        .bind(link_id)
        .execute(pool)
        .await?;

    Ok(())
}

pub async fn create_custom_link(
    pool: &PgPool,
    user_id: i32,
    original_link: &str,
    custom: &str,
) -> Result<(String, bool), sqlx::Error> {
    let link = format!("{}{}", BASE_URL, custom);

    let row: (bool,) = sqlx::query_as("SELECT EXISTS(SELECT 1 FROM links WHERE short_url=$1)")
        .bind(&link)
        .fetch_one(pool)
        .await?;

    if row.0 {
        return Ok(("".to_string(), true));
    }

    sqlx::query(
        "INSERT INTO links (user_id, original_url, short_url, is_custom) VALUES ($1, $2, $3, $4)",
    )
    .bind(user_id)
    .bind(original_link)
    .bind(&link)
    .bind(true)
    .execute(pool)
    .await?;

    Ok((link, false))
}
